### PR TITLE
Upgrade Element UI

### PR DESCRIPTION
**Pull request by OR13**
- add support for JWKs
- fix misc ui issues
- use new keystore
- add qr code scanner for resolving dids from qr codes
- update ui / branding

Screen Shot 2020-01-23 at 4 07 24 PM
Screen Shot 2020-01-23 at 4 08 04 PM

### CODE REVIEW

File: packages/element-app/src/components/Pages/DIDProfilePage/DIDProfilePage.js [hunk 1]
```diff
@@ -2,32 +2,36 @@ import React, { Component } from 'react';
...
+
 import { CreateDefaultDID } from '../../CreateDefaultDID';
 import { DIDDocumentEditorBar } from '../../DIDDocumentEditorBar';
 import { SidetreeOperation } from '../../SidetreeOperation';
 
 export class DIDProfilePage extends Component {
   componentWillMount() {
-    if (!this.props.wallet.data || !this.props.wallet.data.keys) {
+    if (
+      !this.props.keystore.keystore.data ||
```
Comment (gjgd) [inline packages/element-app/src/components/Pages/DIDProfilePage/DIDProfilePage.js]: The page crashes before this line on my machine. Need to add  `!this.props.keystore.keystore ||` as the first condition

File: packages/element-app/src/components/WebKeystore/WebKeystore.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            });
+          }}
+        />
+      </React.Fragment>
+    );
+  }
+}
+
+WebKeystore.propTypes = {
+  keystore: PropTypes.object,
```
Comment (gjgd) [inline packages/element-app/src/components/WebKeystore/WebKeystore.js]: Proptypes are invalid, this causes errors in the console:
Screenshot 2020-01-24 at 13 25 39

File: packages/element-app/src/containers/SnackbarContainer/SnackbarContainer.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import Snackbar from '../../components/Snackbar/Snackbar';
+
+class SnackbarContainer extends React.Component {
+  // componentDidMount() {
```
Comment (gjgd) [inline packages/element-app/src/containers/SnackbarContainer/SnackbarContainer.js]: nit: dead code

File: packages/element-app/src/redux/keystore/index.js [hunk 4]
```diff
@@ -1,9 +1,9 @@
-import * as actions from './actions';
 import reducer from './reducer';
 import container from './container';
 
+// export { reducer, container };
```
Comment (gjgd) [inline packages/element-app/src/redux/keystore/index.js]: nit: dead code

File: packages/element-app/src/redux/keystore/reducer.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { handleActions } from 'redux-actions';
+import { setKeystoreProp } from './actions';
+
+// TODO: UPDATE PLACEHOLDER
```
Comment (gjgd) [inline packages/element-app/src/redux/keystore/reducer.js]: Do we have an issue or a ticket for that TODO?

File: packages/element-app/src/redux/tmui/actions.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import { createAction } from 'redux-actions';
+
+// TODO: UPDATE PLACE HOLDER
```
Comment (gjgd) [inline packages/element-app/src/redux/tmui/actions.js]: Same comment as above

File: packages/element-app/src/redux/tmui/reducer.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { handleActions } from 'redux-actions';
+import { setTmuiProp } from './actions';
+
+// TODO: UPDATE PLACEHOLDER
```
Comment (gjgd) [inline packages/element-app/src/redux/tmui/reducer.js]: same

File: packages/element-lib/src/schema/sidetree/didDocumentModel.json [hunk 8]
```diff
@@ -3,10 +3,12 @@
   "type": "object",
   "properties": {
     "@context": {
-      "type": "string",
-      "required": true,
-      "enum": ["https://w3id.org/did/v1"],
-      "example": ["https://w3id.org/did/v1"]
+      "anyOf": [
```
Comment (gjgd) [inline packages/element-lib/src/schema/sidetree/didDocumentModel.json]: nice

File: packages/element-lib/src/sidetree/op/index.js [hunk 9]
```diff
@@ -117,6 +127,179 @@ const getDeletePayload = async (didUniqueSuffix, recoveryPrivateKey) => {
...
+  const didDocumentModel = walletToInitionalDIDDoc(wall);
+
+  const createPayload = await getCreatePayload(didDocumentModel, primaryKey);
+  const didUniqueSuffix = getDidUniqueSuffix(createPayload);
+
+  const predictedDID = `did:elem:${didUniqueSuffix}`;
+
+  addDIDToWallet(predictedDID, wall);
+
+  // console.log(JSON.stringify(didDocumentModel, null, 2));
```
Comment (gjgd) [inline packages/element-lib/src/sidetree/op/index.js]: nit: cleanup

File: packages/element-app/src/components/Snackbar/MySnackbarContentWrapper.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+import InfoIcon from '@material-ui/icons/Info';
+import CloseIcon from '@material-ui/icons/Close';
+
+import IconButton from '@material-ui/core/IconButton';
+
+import SnackbarContent from '@material-ui/core/SnackbarContent';
+import WarningIcon from '@material-ui/icons/Warning';
+import { makeStyles } from '@material-ui/core/styles';
+
+// import { red, yellow, green, blue } from '@material-ui/core/colors';
```
Comment (gjgd) [inline packages/element-app/src/components/Snackbar/MySnackbarContentWrapper.js]: nit: cleanup

File: packages/element-app/src/components/WebKeystore/ExpansionPanelList.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { withStyles } from '@material-ui/core/styles';
+import ExpansionPanel from '@material-ui/core/ExpansionPanel';
+import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
+import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
+import Typography from '@material-ui/core/Typography';
+import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
+
+// import { grey } from '@material-ui/core/colors';
```
Comment (gjgd) [inline packages/element-app/src/components/WebKeystore/ExpansionPanelList.js]: nit: cleanup

File: packages/element-app/src/components/WebKeystore/ExpansionPanelList.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    width: '100%',
+    marginBottom: '4px',
+  },
+  heading: {
+    fontSize: theme.typography.pxToRem(15),
+    fontWeight: theme.typography.fontWeightRegular,
+    wordBreak: 'break-all',
+  },
+  panel: {
+    // backgroundColor: grey[200],
```
Comment (gjgd) [inline packages/element-app/src/components/WebKeystore/ExpansionPanelList.js]: nit: cleanup

File: packages/element-app/src/components/WebKeystore/KeystoreEditor.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+// eslint-disable-next-line
```
Comment (gjgd) [inline packages/element-app/src/components/WebKeystore/KeystoreEditor.js]: the warning says brace should be listed in the project's dependencies. Any reasons why not doing it?

Comment (gjgd) [reply packages/element-app/src/components/WebKeystore/KeystoreEditor.js]: Also, looks like brace is not used in this file

File: packages/element-app/src/redux/sidetreeWallet/handlers.js [hunk 14]
```diff
@@ -1,6 +1,8 @@
 import { withHandlers } from 'recompose';
 import { func, op } from '@transmute/element-lib';
 
+const DIDWallet = require('@transmute/did-wallet');
```
Comment (gjgd) [inline packages/element-app/src/redux/sidetreeWallet/handlers.js]: let's use the `import` syntax for consistency

File: packages/element-app/package.json [hunk 15]
```diff
@@ -5,25 +5,28 @@
...
     "@material-ui/icons": "^3.0.2",
-    "@transmute/did-wallet": "0.0.0-5",
+    "@transmute/did-wallet": "0.0.0-6",
     "@transmute/element-lib": "^0.4.3",
     "axios": "^0.18.0",
     "base64url": "^3.0.1",
     "classnames": "^2.2.6",
+    "clsx": "^1.0.4",
     "connected-react-router": "^6.2.1",
+    "crypto-ld": "^3.7.0",
```
Comment (gjgd) [inline packages/element-app/package.json]: unused

File: packages/element-app/package.json [hunk 16]
```diff
@@ -5,25 +5,28 @@
...
+    "did-method-key": "^0.3.0",
```
Comment (gjgd) [inline packages/element-app/package.json]: unused, we should be careful about adding more crypto dependencies to our already huge app bundle

File: packages/element-lib/src/sidetree/op/index.js [hunk 17]
```diff
@@ -117,6 +127,179 @@ const getDeletePayload = async (didUniqueSuffix, recoveryPrivateKey) => {
...
+            usage:
+              walletKey.tags[1] &&
+              walletKey.tags[1].split('#').pop() === 'recovery'
+                ? 'recovery'
+                : 'signing',
+            [walletKey.didPublicKeyEncoding]: walletKey.publicKey,
+          });
+        }
+
+        commonVerificationMethods.push(`#${walletKey.kid}`);
```
Comment (gjgd) [inline packages/element-lib/src/sidetree/op/index.js]: This is not correct. See this did:
```json
{
  "authentication": [
    "#a4VSF7tSJ7VKSOwkI98m3OHz0OHGIdS5DB_b4fj4gGM",
    "#NFqToUusDIZ3dv1I0xWjWXQFFGqstQK2hntYISIVJpo",
    "#_vLQrqsD3zuRvkVdM0du2Q-CaZPhWWsgxjqO8VPmpoE",
    "#AwE8YZ7HcJ2zORh_XbrDPTFlHTAeNZkLKSrnxYKLnf0"
  ],
  "capabilityDelegation": [
    "#a4VSF7tSJ7VKSOwkI98m3OHz0OHGIdS5DB_b4fj4gGM",
    "#NFqToUusDIZ3dv1I0xWjWXQFFGqstQK2hntYISIVJpo",
    "#_vLQrqsD3zuRvkVdM0du2Q-CaZPhWWsgxjqO8VPmpoE",
    "#AwE8YZ7HcJ2zORh_XbrDPTFlHTAeNZkLKSrnxYKLnf0"
  ],
  "keyAgreement": [
    {
      "type": "X25519KeyAgreementKey2019",
      "publicKeyBase58": "3TkMWjeZvnFtJ1wQZTaSWSruGqsU5GZmRLdJ1YcqKzSp",
      "id": "#keyAgreement",
      "usage": "signing",
      "controller": "did:elem:EiBiiyNodE1Kb3ozmMXp1o9pe_s6Yg-UYUAC4Vh_ZUy6mg"
    }
  ],
  "assertionMethod": [
    "#a4VSF7tSJ7VKSOwkI98m3OHz0OHGIdS5DB_b4fj4gGM",
    "#NFqToUusDIZ3dv1I0xWjWXQFFGqstQK2hntYISIVJpo",
    "#_vLQrqsD3zuRvkVdM0du2Q-CaZPhWWsgxjqO8VPmpoE",
    "#AwE8YZ7HcJ2zORh_XbrDPTFlHTAeNZkLKSrnxYKLnf0"
  ],
  "@context": [
    "https://www.w3.org/ns/did/v1",
    "https://docs.element-did.com/contexts/sidetree/sidetree-v0.1.jsonld"
  ],
  "publicKey": [
    {
      "publicKeyJwk": {
        "x": "HfPpOQ9Qvs3OaQG4-mxpAkAv0HAL2IB_xQP_SRvZuEA",
        "kty": "EC",
        "crv": "secp256k1",
        "y": "-luk_H74kMnty-ko5nIfMqbiw9BxbcRGFijJH4GWWmQ",
        "kid": "a4VSF7tSJ7VKSOwkI98m3OHz0OHGIdS5DB_b4fj4gGM"
      },
      "type": "EcdsaSecp256k1VerificationKey2019",
      "id": "#a4VSF7tSJ7VKSOwkI98m3OHz0OHGIdS5DB_b4fj4gGM",
      "usage": "signing",
      "controller": "did:elem:EiBiiyNodE1Kb3ozmMXp1o9pe_s6Yg-UYUAC4Vh_ZUy6mg"
    },
    {
      "type": "EcdsaSecp256k1VerificationKey2019",
      "id": "#primary",
      "usage": "signing",
      "publicKeyHex": "021df3e9390f50becdce6901b8fa6c6902402fd0700bd8807fc503ff491bd9b840",
      "controller": "did:elem:EiBiiyNodE1Kb3ozmMXp1o9pe_s6Yg-UYUAC4Vh_ZUy6mg"
    },
    {
      "publicKeyHex": "0305347588cb15c7d7cb10ffd94ec0abc1f98428209447b6644a7bced1197ff603",
      "type": "EcdsaSecp256k1VerificationKey2019",
      "id": "#recovery",
      "usage": "recovery",
      "controller": "did:elem:EiBiiyNodE1Kb3ozmMXp1o9pe_s6Yg-UYUAC4Vh_ZUy6mg"
    },
    {
      "type": "Ed25519VerificationKey2018",
      "publicKeyBase58": "CzcfYNghXxbjaYNfqvf8wPMS4va3cLHGXaoUdxZd1DWC",
      "id": "#AwE8YZ7HcJ2zORh_XbrDPTFlHTAeNZkLKSrnxYKLnf0",
      "usage": "signing",
      "controller": "did:elem:EiBiiyNodE1Kb3ozmMXp1o9pe_s6Yg-UYUAC4Vh_ZUy6mg"
    }
  ],
  "capabilityInvocation": [
    "#a4VSF7tSJ7VKSOwkI98m3OHz0OHGIdS5DB_b4fj4gGM",
    "#NFqToUusDIZ3dv1I0xWjWXQFFGqstQK2hntYISIVJpo",
    "#_vLQrqsD3zuRvkVdM0du2Q-CaZPhWWsgxjqO8VPmpoE",
    "#AwE8YZ7HcJ2zORh_XbrDPTFlHTAeNZkLKSrnxYKLnf0"
  ],
  "id": "did:elem:EiBiiyNodE1Kb3ozmMXp1o9pe_s6Yg-UYUAC4Vh_ZUy6mg"
}
```

only the first capabilityInvocation: #a4VSF7tSJ7VKSOwkI98m3OHz0OHGIdS5DB_b4fj4gGM is a valid reference to a key in the document

File: packages/element-app/src/components/Snackbar/index.js [hunk 18]
```diff
@@ -1,5 +0,0 @@
-import Snackbar from './Snackbar';
```
Comment (gjgd) [inline packages/element-app/src/components/Snackbar/index.js]: Not sure what the best place to put this is, but the new snackbar logic causes some prop validation to fail. For example in FullNodeExplorerTransactionPage:
Screenshot 2020-01-24 at 14 08 08

File: packages/element-app/package.json [hunk 19]
```diff
@@ -5,25 +5,28 @@
   "homepage": "https://element-did.com",
```
Comment (gjgd) [inline packages/element-app/package.json]: Adding the BIP39 key or the keyagreement key does not work.

 Let's filter out keys that cannot be added in this dialog. (Btw why do we have this type of key?)

Screenshot 2020-01-24 at 14 26 28

File: packages/element-lib/src/sidetree/op/index.js [hunk 20]
```diff
@@ -117,6 +127,179 @@ const getDeletePayload = async (didUniqueSuffix, recoveryPrivateKey) => {
...
+  const x25519Key = await elementCrypto.ed25519.X25519KeyPair.fromEdKeyPair({
+    publicKeyBase58: ed25519Key.publicKey,
+    privateKeyBase58: ed25519Key.privateKey,
+  });
+
+  const wall = DIDWallet.create();
+
+  const notes = 'generated in element-lib.';
+
+  wall.addKey({
```
Comment (gjgd) [inline packages/element-lib/src/sidetree/op/index.js]: why add the mnemonic?

File: packages/element-app/src/components/WebKeystore/KeystoreCard.js [hunk 21]
```diff
@@ -0,0 +1,372 @@
...
+                                  // eslint-disable-next-line no-alert
+                                  alert('Copied public key to clipboard');
+                                }}
+                              >
+                                <Button
+                                  style={{ marginTop: '16px' }}
+                                  fullWidth
+                                  variant="contained"
+                                >
+                                  Copy Public Key
```
Comment (gjgd) [inline packages/element-app/src/components/WebKeystore/KeystoreCard.js]: Not working for bip39 mnemonic

File: packages/element-app/src/components/Pages/DIDProfilePage/DIDProfilePage.js (hunk 1 above)
Comment (OR13) [reply packages/element-app/src/components/Pages/DIDProfilePage/DIDProfilePage.js]: excellent catch

File: packages/element-app/src/components/WebKeystore/ExpansionPanelList.js (hunk 11 above)
Comment (OR13) [reply packages/element-app/src/components/WebKeystore/ExpansionPanelList.js]: leaving this for non branded case... this should really be pulled from theme, but we will wait solve that in a proper material package.

File: packages/element-app/src/components/WebKeystore/KeystoreEditor.js (hunk 13 above)
Comment (OR13) [reply packages/element-app/src/components/WebKeystore/KeystoreEditor.js]: this is a weired issue with ace... its required.

Comment (OR13) [reply packages/element-app/src/components/WebKeystore/KeystoreEditor.js]: has to do with themes...

File: packages/element-app/package.json (hunk 16 above)
Comment (OR13) [reply packages/element-app/package.json]: great catch, I had to add these as part of the keystore upgrade, forgot to remove them.

File: packages/element-lib/src/sidetree/op/index.js (hunk 17 above)
Comment (OR13) [reply packages/element-lib/src/sidetree/op/index.js]: great catch...

File: packages/element-app/src/components/Snackbar/index.js (hunk 18 above)
Comment (OR13) [reply packages/element-app/src/components/Snackbar/index.js]: great catch.

Comment (OR13) [reply packages/element-app/src/components/Snackbar/index.js]: Solved by making sure not to set a prop to undefined...

```
  componentDidMount() {
    this.setState({
      expanded: this.props.expanded || false,
    });
  }
```

File: packages/element-lib/src/sidetree/op/index.js (hunk 20 above)
Comment (OR13) [reply packages/element-lib/src/sidetree/op/index.js]: we use it to generate all keys using the MKS system... if we want to be able to get back to all keys in an indexable manner, we need to add it.

See this issue for more context: https://github.com/w3c-ccg/webkms/issues/2

File: packages/element-app/src/components/WebKeystore/KeystoreCard.js (hunk 21 above)
Comment (OR13) [reply packages/element-app/src/components/WebKeystore/KeystoreCard.js]: I made it not possible to copy it, you can see it in the editor though.

File: packages/element-app/src/components/Snackbar/MySnackbarContentWrapper.js (hunk 10 above)
Comment (OR13) [reply packages/element-app/src/components/Snackbar/MySnackbarContentWrapper.js]: I opted to comment this for the case where we support non branded options.

File: packages/element-app/src/components/WebKeystore/ExpansionPanelList.js (hunk 11 above)
Comment (OR13) [reply packages/element-app/src/components/WebKeystore/ExpansionPanelList.js]: this comment belonged on the comment above :(

**Pull request merged**